### PR TITLE
Formulate elliptic systems in terms of fluxes and sources

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Refactors the elliptic code to a formulation that resembles closely the
flux-conservative form of evolution systems. The system just computes
the fluxes and sources and no longer has to implement the derivative
operator or the normal-dot-fluxes. Instead, they use code shared
between systems, so adding the Elasticity and XCTS systems (next PRs)
is straight-forward.

**Notes to reviewers:**
- This PR changes a lot of lines, but most commits should be straight-forward to review since they do simple renames.
- I tried to split this PR into many small commits, so it's best to step through the commit history when reviewing.
- <s>I could defer the last 4 commits to a follow-up PR if the reviewers prefer. They should be quite easy to review, so it's probably more effective to keep them here. They actually add some important tests for the code in this PR.</s> Deferred this as requested.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsdeppe** [comment]
I would recommend splitting this into 3 or 4 PRs if possible. Ideally PRs should be 400 lines or less. I realize that's sometimes not possible, but considering you have 13 commits it seems like it should be possible to at least downsize this a bit.

**Comment by nilsvu** [comment]
@nilsdeppe Ok, I dropped the last 4 commits. Now it's a bunch of simple preparation-commits leading up to "Formulate elliptic system in terms of fluxes and sources". I think I have now moved everything I could out of that commit.

File: src/Elliptic/FirstOrderOperator.hpp
Comment (kidder) [inline src/Elliptic/FirstOrderOperator.hpp]: `sources - div_fluxes`

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (kidder) [inline src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp]: When are you going to `CHECK_THIS` ?

**Comment by kidder** [comment]
@geoffrey4444 please review this

**Comment by nilsvu** [comment]
@kidder fixed the issues you commented on and accidentally squashed those in, sorry 😅

File: src/Elliptic/Actions/InitializeSystem.hpp
Comment (geoffrey4444) [inline src/Elliptic/Actions/InitializeSystem.hpp]: Safer to initialize to signaling nan?

File: src/Elliptic/Actions/InitializeSystem.hpp
Comment (geoffrey4444) [inline src/Elliptic/Actions/InitializeSystem.hpp:129]: Just to help me understand, where do you actually specify what the source term is (e.g., the charge density for Poisson)?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/Test_ConstantDensityStar.cpp
Comment (geoffrey4444) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/Test_ConstantDensityStar.cpp]: Why not FixedSource?

File: src/Elliptic/FirstOrderComputeTags.hpp
Comment (geoffrey4444) [inline src/Elliptic/FirstOrderComputeTags.hpp:24]: To help me learn, what do lines 18-24 do here?

File: tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp
Comment (geoffrey4444) [inline tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp]: Maybe do something different for the sources than the tags, like multiply by sqr(an_argument), to make the test more generic?

File: tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp
Comment (geoffrey4444) [inline tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp:106]: Maybe it's too trivial to be worth it, but maybe check that the terms you didn't set are actually zero?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
Started reviewing. Looked at commits 85e0312 through 2b4cb75. Just some minor comments and some questions to help me learn

File: src/Elliptic/FirstOrderOperator.hpp
Comment (geoffrey4444) [inline src/Elliptic/FirstOrderOperator.hpp]: What is a StepPrefixTag? What possible values can it take? Please specify here or point to relevant docs

File: tests/Unit/Elliptic/Test_FirstOrderOperator.cpp
Comment (geoffrey4444) [inline tests/Unit/Elliptic/Test_FirstOrderOperator.cpp]: Do you have to specify the `name()` if it's the same as the struct's name?

File: tests/Unit/Elliptic/Test_FirstOrderOperator.cpp
Comment (geoffrey4444) [inline tests/Unit/Elliptic/Test_FirstOrderOperator.cpp]: Shouldn't this name be `AnotherField`?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Moustache.py
Comment (geoffrey4444) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Moustache.py:22]: Just a nitpick: I'm unclear why the indentation changed here

File: src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp
Comment (geoffrey4444) [inline src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp]: Can you please briefly specify what will need to be extended to implement other BCs than homogeneous Dirichlet?

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (geoffrey4444) [inline src/Elliptic/Systems/Poisson/Equations.cpp:38]: How would you handle an elliptic equation whose fields include a vector field? I'm just a little confused why this is called `field` instead of `scalar_field`

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (geoffrey4444) [inline src/Elliptic/Systems/Poisson/Equations.cpp]: Can you please explain why this change? I thought in a previous commit in this PR you changed the gradient to a lower index. In terms of math, I would have thought the gradient of a field would be a convector

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (geoffrey4444) [inline src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:37]: Just a nitpick, but I prefer \nabla^2 to \Delta for the Laplacian operator

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (geoffrey4444) [inline src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp]: \f$\gamma_{ij}\f$

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (geoffrey4444) [inline src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:80]: Will preconditioning be more complicated for this first-order formulation?

**Review by geoffrey4444 — CHANGES_REQUESTED** [review]
More comments...still need to finish looking through Test_InitializeFluxes.cpp and Test_Equations.cpp in 579b2de and then the bottom 3 commits

File: src/Elliptic/Actions/InitializeSystem.hpp
Comment (nilsvu) [reply src/Elliptic/Actions/InitializeSystem.hpp]: This initializes the auxiliary field sources to `0.`

File: src/Elliptic/Actions/InitializeSystem.hpp
Comment (nilsvu) [reply src/Elliptic/Actions/InitializeSystem.hpp:129]: That's what this bit of code is about. The "fixed sources" (i.e. the `f(x)` in `-Laplace(u(x))=f(x)`) are being retrieved from the analytic solution. You can find examples in `Poisson::Solutions`.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/Test_ConstantDensityStar.cpp
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/Test_ConstantDensityStar.cpp]: This indeed changes to `FixedSource` once the tag is added in one of the following commits.

File: src/Elliptic/FirstOrderComputeTags.hpp
Comment (nilsvu) [reply src/Elliptic/FirstOrderComputeTags.hpp:24]: - I need to expand the tags typelists into parameter packs, which is why this forward declaration takes typelists, and the template specialization below extracts the tags they contain as a parameter pack.
- I use these defaulted template parameters to retrieve the typelists from the `System` (and they can also be overridden if other tags should be used).

File: tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp
Comment (nilsvu) [reply tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp]: good idea 👍

File: tests/Unit/Elliptic/Test_FirstOrderOperator.cpp
Comment (nilsvu) [reply tests/Unit/Elliptic/Test_FirstOrderOperator.cpp]: DataBox tag names don't default to the class names (yet).

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Moustache.py
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Moustache.py:22]: `autopep8` probably did that. I can revert the change

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/Equations.cpp:38]: This is the implementation for the Poisson system, so I'm working with the (scalar) `Poisson::Tags::Field` and its gradient. An example for an elliptic system that uses a vector field is the Elasticity system that I'll add once these preparatory PRs are in. It implements its own fluxes and sources for the `Elasticity::Tags::Displacement` vector field and the `Elasticity::Tags::Strain` symmetric rank-2 tensor.

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/Equations.cpp]: This quantity is now the principal symbol applied to the gradient, i.e. for the Euclidean Poisson system it's just the gradient with its index raised with the flat metric. I didn't change the variable name because for the Poisson system it's still the same values and I want to keep changes to the flux to a minimum. That's because I'll delete this Poisson-specific implementation of the flux in #1671 and replace it with a generic internal penalty flux for elliptic systems.

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:80]: No, preconditioning doesn't care about the formulation :) It's an operation on the linear operator and agnostic to how it's built.

File: tests/Unit/Elliptic/Test_FirstOrderOperator.cpp
Comment (wthrowe) [reply tests/Unit/Elliptic/Test_FirstOrderOperator.cpp]: Yes they do.  That feature was added with PDALs (#1535).  (Some things still try to call `Tag::name()` directly instead of `db::tag_name<Tag>()`, but those are easy-to-fix bugs.)

Comment (nilsvu) [reply tests/Unit/Elliptic/Test_FirstOrderOperator.cpp]: You're right of course. I didn't realize `db::tag_name` had that functionality 👍

File: tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp
Comment (nilsvu) [reply tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp:106]: good point, added a check

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Moustache.py
Comment (nilsvu) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Moustache.py:22]: actually, since `autopep8` wants this formatting, I'll just keep it like this if that's ok

File: src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp]: This is actually valid for any boundary conditions, adjusted the comment. It's the actions in Elliptic/DG that need to be changed to support other boundary conditions (they are not affected by this PR). They can currently handle inhomogeneous Dirichlet b.c.

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:37]: It's `\Delta` in a lot of places in the documentation already and I don't see a strong reason to change that, in particular because also much of the (NR) literature uses `\Delta`. It's a matter of personal preference I guess.

**Comment by nilsvu** [comment]
@geoffrey4444 I pushed fixups and responded to your comments. Please take a look :)

File: src/Elliptic/FirstOrderOperator.hpp
Comment (nilsvu) [reply src/Elliptic/FirstOrderOperator.hpp]: Added more details to the docs

**Comment by kidder** [comment]
@geoffrey4444 please look at the fixups

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (geoffrey4444) [reply src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:37]: I don't feel strongly about it...I'm ok leaving it as is.

**Review by geoffrey4444 — DISMISSED** [review]

File: src/Elliptic/FirstOrderOperator.hpp
Comment (kidder) [reply src/Elliptic/FirstOrderOperator.hpp]: I still don't understand what a StepPrefixTag is.  Can you point to an equation illustrating it?

**Comment by kidder** [comment]
@wthrowe can you look over the last two commits of this PR today or tomorrow?  I went over the first 9 commits and have no requested changes.  And Geoffrey has already approved.

File: src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp:71]: I just want to verify that this slicing is the correct behavior, and not a hack like what we have in the evolution code.  If it is not correct please file an issue or comment on the existing boundary condition issue or something.

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (wthrowe) [inline src/Elliptic/Systems/Poisson/Equations.cpp]: Only do a resize if necessary.

File: src/Elliptic/Systems/Poisson/Equations.hpp
Comment (wthrowe) [inline src/Elliptic/Systems/Poisson/Equations.hpp]: Only resize if necessary.

File: src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp
Comment (wthrowe) [inline src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp]: Rendering error.

**Review by wthrowe — CHANGES_REQUESTED** [review]
I reviewed the topologically last two commits:  "Formulate elliptic system in terms of fluxes and sources" and "Use Poisson field gradient for auxiliary variable".  (I had no comments on the latter of those two.)  If you wanted me to look at the (author-)temporally last two commits instead, let me know.

File: src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp:71]: It‘s the right behavior on Dirichlet boundaries, which is all that‘s currently implemented. I‘ll have to revisit this when we add more general boundary conditions.

**Comment by kidder** [comment]
@wthrowe sorry for not being clear, please also look at "Add first-order elliptic operator"

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/Equations.cpp]: This sets all non-diagonal components to zero. Is there a better way?

File: src/Elliptic/Systems/Poisson/Equations.hpp
Comment (nilsvu) [reply src/Elliptic/Systems/Poisson/Equations.hpp]: same here, it zeroes the source

File: src/Elliptic/FirstOrderOperator.hpp
Comment (nilsvu) [reply src/Elliptic/FirstOrderOperator.hpp]: pushed a fixup to address this, please take a look

File: src/Elliptic/FirstOrderOperator.hpp
Comment (wthrowe) [inline src/Elliptic/FirstOrderOperator.hpp]: Missing `\f`s.

File: src/Elliptic/FirstOrderOperator.hpp
Comment (wthrowe) [inline src/Elliptic/FirstOrderOperator.hpp:55]: The documentation for this struct does not seem to be describing the correct thing.  It keeps talking about actions and describes an action interface, but this isn't an action.

File: src/Elliptic/Systems/Poisson/Equations.cpp
Comment (wthrowe) [reply src/Elliptic/Systems/Poisson/Equations.cpp]: I see `std::fill` used in some places, and I can't think of anything better.

File: src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp
Comment (wthrowe) [reply src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp:71]: OK.  Thanks for confirming.

File: src/Elliptic/Systems/Poisson/Equations.hpp
Comment (wthrowe) [reply src/Elliptic/Systems/Poisson/Equations.hpp]: This one can just use `get(...) = 0.` since it's a scalar.  (After resizing if necessary.)

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Elliptic/FirstOrderOperator.hpp
Comment (nilsvu) [reply src/Elliptic/FirstOrderOperator.hpp:55]: made some changes to the docs. Lacking a better name I refer to it as an _invokable_ now. I don't know if we have decided on a name for these classes that conform to `mutate_apply`.

**Comment by nilsvu** [comment]
@wthrowe @kidder please check the fixups

File: src/Elliptic/FirstOrderOperator.hpp
Comment (wthrowe) [inline src/Elliptic/FirstOrderOperator.hpp]: Something is wrong with the wording here.

File: src/Elliptic/FirstOrderOperator.hpp
Comment (wthrowe) [inline src/Elliptic/FirstOrderOperator.hpp]: Not an action.

**Comment by nilsvu** [comment]
@wthrowe squashed corrections for your last two comments into the `fixup FirstOrderOperator`. Are you and @kidder fine with squashing the fixups?

**Comment by wthrowe** [comment]
I am fine with this.

**Comment by kidder** [comment]
me too

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe this should be ready to merge

**Comment by nilsvu** [comment]
@geoffrey4444 please re-approve as well when you're happy with this.

**Review by wthrowe — APPROVED** [review]
Waiting for approval to merge by @geoffrey4444 per author request.

**Review by geoffrey4444 — APPROVED** [review]

**Comment by nilsvu** [comment]
Thanks for reviewing @geoffrey4444 @wthrowe @kidder! The story continues in #1671 where I add a generic numerical flux for elliptic systems. Could you be the reviewers for that one as well?

**Pull request merged**
